File: js/test/config.test.ts
import { expect, test, vi } from "vitest";
import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { tmpdir } from "node:os";
import path from "node:path";
import { configFilePath, getProfile } from "../src/config";
import { DEFAULT_SANDBOX_CHANNEL_IDLE_TIMEOUT_MS } from "../src/config";

const maxThrottleWaitCases = [
  { envVal: "10", expected: 10 },
  { envVal: "0", expected: 0 },
  { envVal: "3600", expected: 3600 },
];

for (const { envVal, expected } of maxThrottleWaitCases) {
  test(`GetProfile_MaxThrottleWaitParsing/${envVal}`, () => {
    vi.stubEnv("MODAL_MAX_THROTTLE_WAIT", envVal);
    const profile = getProfile();
    expect(profile.maxThrottleWaitSecs).toBe(expected);
    vi.unstubAllEnvs();
  });
}

test("GetProfile_MaxThrottleWaitInvalidValue", () => {
  vi.stubEnv("MODAL_MAX_THROTTLE_WAIT", "not-a-number");
  const profile = getProfile();
  expect(profile.maxThrottleWaitSecs).toBeUndefined();
  vi.unstubAllEnvs();
});

const sandboxV2Cases = [
  { envVal: undefined, expected: false },
  { envVal: "", expected: false },
  { envVal: "0", expected: false },
  { envVal: "false", expected: false },
  { envVal: "False", expected: false },
  { envVal: "1", expected: true },
  { envVal: "true", expected: true },
  { envVal: "yes", expected: true },
];

for (const { envVal, expected } of sandboxV2Cases) {
  test(`GetProfile_SandboxV2Parsing/${JSON.stringify(envVal)}`, () => {
    vi.stubEnv("MODAL_SANDBOX_V2", envVal);
    const profile = getProfile();
    expect(profile.sandboxV2).toBe(expected);
    vi.unstubAllEnvs();
  });
}

test("GetProfile_SandboxV2FromConfig", async () => {
  const configDir = mkdtempSync(path.join(tmpdir(), "modal-js-config-"));
  const configPath = path.join(configDir, ".modal.toml");
  writeFileSync(
    configPath,
    `
[v2-profile]
sandbox_v2 = true
image_builder_version = "2024.10"
`,
  );
  vi.stubEnv("MODAL_CONFIG_PATH", configPath);
  vi.stubEnv("MODAL_SANDBOX_V2", undefined);
  vi.stubEnv("MODAL_IMAGE_BUILDER_VERSION", undefined);
  vi.resetModules();

  try {
    const { getProfile: getProfileFromConfig } = await import("../src/config");
    const profile = getProfileFromConfig("v2-profile");
    expect(profile.sandboxV2).toBe(true);
    expect(profile.imageBuilderVersion).toBe("2024.10");
  } finally {
    vi.unstubAllEnvs();
    vi.resetModules();
    rmSync(configDir, { recursive: true });
  }
});

test("GetProfile_SandboxV2QuotedValuesInConfig", async () => {
  const configDir = mkdtempSync(path.join(tmpdir(), "modal-js-config-"));
  const configPath = path.join(configDir, ".modal.toml");
  writeFileSync(
    configPath,
    `
[quoted-false]
sandbox_v2 = "false"

[quoted-true]
sandbox_v2 = "true"
`,
  );
  vi.stubEnv("MODAL_CONFIG_PATH", configPath);
  vi.stubEnv("MODAL_SANDBOX_V2", undefined);
  vi.resetModules();

  try {
    const { getProfile: getProfileFromConfig } = await import("../src/config");
    expect(getProfileFromConfig("quoted-false").sandboxV2).toBe(false);
    expect(getProfileFromConfig("quoted-true").sandboxV2).toBe(true);
  } finally {
    vi.unstubAllEnvs();
    vi.resetModules();
    rmSync(configDir, { recursive: true });
  }
});

test("GetProfile_SandboxV2EnvOverridesConfig", async () => {
  const configDir = mkdtempSync(path.join(tmpdir(), "modal-js-config-"));
  const configPath = path.join(configDir, ".modal.toml");
  writeFileSync(
    configPath,
    `
[v2-profile]
sandbox_v2 = true
`,
  );
  vi.stubEnv("MODAL_CONFIG_PATH", configPath);
  vi.stubEnv("MODAL_SANDBOX_V2", "0");
  vi.resetModules();

  try {
    const { getProfile: getProfileFromConfig } = await import("../src/config");
    const profile = getProfileFromConfig("v2-profile");
    expect(profile.sandboxV2).toBe(false);
  } finally {
    vi.unstubAllEnvs();
    vi.resetModules();
    rmSync(configDir, { recursive: true });
  }
});

test("GetProfile_OAuthCredentials", () => {
  vi.stubEnv("MODAL_OAUTH_REFRESH_TOKEN", "refresh-token");
  vi.stubEnv("MODAL_OAUTH_CLIENT_ID", "oc-client-id");
  vi.stubEnv("MODAL_OAUTH_CLIENT_SECRET", "ov-client-secret");

  const profile = getProfile();
  expect(profile.oauthRefreshToken).toBe("refresh-token");
  expect(profile.oauthClientId).toBe("oc-client-id");
  expect(profile.oauthClientSecret).toBe("ov-client-secret");
  vi.unstubAllEnvs();
});

test("GetProfile_OAuthCredentialsFromConfig", async () => {
  const configDir = mkdtempSync(path.join(tmpdir(), "modal-js-config-"));
  const configPath = path.join(configDir, ".modal.toml");
  writeFileSync(
    configPath,
    `
[oauth-profile]
oauth_refresh_token = "refresh-token"
oauth_client_id = "oc-client-id"
oauth_client_secret = "ov-client-secret"
`,
  );
  vi.stubEnv("MODAL_CONFIG_PATH", configPath);
  vi.stubEnv("MODAL_OAUTH_REFRESH_TOKEN", undefined);
  vi.stubEnv("MODAL_OAUTH_CLIENT_ID", undefined);
  vi.stubEnv("MODAL_OAUTH_CLIENT_SECRET", undefined);
  vi.resetModules();

  try {
    const { getProfile: getProfileFromConfig } = await import("../src/config");
    const profile = getProfileFromConfig("oauth-profile");
    expect(profile.oauthRefreshToken).toBe("refresh-token");
    expect(profile.oauthClientId).toBe("oc-client-id");
    expect(profile.oauthClientSecret).toBe("ov-client-secret");
  } finally {
    vi.unstubAllEnvs();
    vi.resetModules();
    rmSync(configDir, { recursive: true });
  }
});

test("GetConfigPath_WithEnvVar", () => {
  const customPath = "/custom/path/to/config.toml";
  vi.stubEnv("MODAL_CONFIG_PATH", customPath);

  const result = configFilePath();
  expect(result).toBe(customPath);

  vi.unstubAllEnvs();
});

test("GetConfigPath_WithoutEnvVar", () => {
  vi.stubEnv("MODAL_CONFIG_PATH", undefined);

  const result = configFilePath();
  const expectedPath = path.join(homedir(), ".modal.toml");
  expect(result).toBe(expectedPath);

  vi.unstubAllEnvs();
});

// A value that parses as a number is not necessarily a timeout. Infinity is
// not one, and setTimeout treats anything past 2^31-1 ms as zero, so a value
// that large would fire at once rather than never.
const notATimeout = ["Infinity", "-Infinity", "NaN", "1e30", "-1", "nonsense"];

for (const value of notATimeout) {
  test(`GetProfile_ChannelIdleTimeoutRejects/${value}`, () => {
    vi.stubEnv("MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", value);
    const profile = getProfile();
    expect(profile.sandboxChannelIdleTimeoutMs).toBe(
      DEFAULT_SANDBOX_CHANNEL_IDLE_TIMEOUT_MS,
    );
    vi.unstubAllEnvs();
  });
}

// Values that are timeouts still get through, zero included: it turns the
// release off rather than releasing at once.
test("GetProfile_IdleTimeoutAcceptsSeconds", () => {
  vi.stubEnv("MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", "0");
  const profile = getProfile();
  expect(profile.sandboxChannelIdleTimeoutMs).toBe(0);
  vi.unstubAllEnvs();
});

// A positive timeout too short to round to a millisecond must not read as zero,
// which is how the release is turned off - that would invert what was asked for.
test("GetProfile_IdleTimeoutKeepsAShortTimeoutPositive", () => {
  vi.stubEnv("MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", "0.0004");
  const profile = getProfile();
  expect(profile.sandboxChannelIdleTimeoutMs).toBeGreaterThan(0);
  vi.unstubAllEnvs();
});

File: go/config_test.go
package modal

import (
	"os"
	"path/filepath"
	"strconv"
	"testing"
	"time"

	"github.com/onsi/gomega"
)

func TestGetConfigPath_WithEnvVar(t *testing.T) {
	g := gomega.NewWithT(t)

	customPath := "/custom/path/to/config.toml"
	t.Setenv("MODAL_CONFIG_PATH", customPath)

	path, err := configFilePath()
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(path).Should(gomega.Equal(customPath))
}

func TestGetConfigPath_WithoutEnvVar(t *testing.T) {
	g := gomega.NewWithT(t)

	t.Setenv("MODAL_CONFIG_PATH", "")

	path, err := configFilePath()
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	home, _ := os.UserHomeDir()
	expectedPath := filepath.Join(home, ".modal.toml")
	g.Expect(path).Should(gomega.Equal(expectedPath))
}

func TestGetProfile_MaxThrottleWaitParsing(t *testing.T) {
	cases := []struct {
		envVal   string
		expected *time.Duration
	}{
		{"10", durationPtr(10 * time.Second)},
		{"0", durationPtr(0)},
		{"3600", durationPtr(3600 * time.Second)},
	}

	for _, tc := range cases {
		t.Run(tc.envVal, func(t *testing.T) {
			g := gomega.NewWithT(t)
			t.Setenv("MODAL_MAX_THROTTLE_WAIT", tc.envVal)
			profile := getProfile("", config{})
			g.Expect(profile.MaxThrottleWait).NotTo(gomega.BeNil())
			g.Expect(*profile.MaxThrottleWait).To(gomega.Equal(*tc.expected))
		})
	}
}

func durationPtr(d time.Duration) *time.Duration { return &d }

func TestGetProfile_MaxThrottleWaitInvalidValue(t *testing.T) {
	g := gomega.NewWithT(t)

	t.Setenv("MODAL_MAX_THROTTLE_WAIT", "not-a-number")

	profile := getProfile("", config{})
	g.Expect(profile.MaxThrottleWait).To(gomega.BeNil())
}

func TestGetProfile_OAuthCredentials(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_OAUTH_REFRESH_TOKEN", "refresh-token")
	t.Setenv("MODAL_OAUTH_CLIENT_ID", "oc-client-id")
	t.Setenv("MODAL_OAUTH_CLIENT_SECRET", "ov-client-secret")

	profile := getProfile("", config{})
	g.Expect(profile.OAuthRefreshToken).To(gomega.Equal("refresh-token"))
	g.Expect(profile.OAuthClientID).To(gomega.Equal("oc-client-id"))
	g.Expect(profile.OAuthClientSecret).To(gomega.Equal("ov-client-secret"))
}

func TestGetProfile_OAuthJWTKey(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_OAUTH_REFRESH_TOKEN", "refresh-token")
	t.Setenv("MODAL_OAUTH_CLIENT_ID", "oc-client-id")
	t.Setenv("MODAL_OAUTH_JWT_KEY", "private-key")

	profile := getProfile("", config{})
	g.Expect(profile.OAuthRefreshToken).To(gomega.Equal("refresh-token"))
	g.Expect(profile.OAuthClientID).To(gomega.Equal("oc-client-id"))
	g.Expect(profile.OAuthJWTKey).To(gomega.Equal("private-key"))
	g.Expect(profile.OAuthClientSecret).To(gomega.BeEmpty())
}

func TestGetProfile_OAuthJWTEnvClearsFileSecret(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_TOKEN_ID", "")
	t.Setenv("MODAL_TOKEN_SECRET", "")
	t.Setenv("MODAL_OAUTH_CLIENT_SECRET", "")
	t.Setenv("MODAL_OAUTH_JWT_KEY", "env-jwt-key")

	profile := getProfile("oauth-profile", config{
		"oauth-profile": rawProfile{
			OAuthRefreshToken: "refresh-token",
			OAuthClientID:     "oc-client-id",
			OAuthClientSecret: "ov-file-secret",
		},
	})
	g.Expect(profile.OAuthJWTKey).To(gomega.Equal("env-jwt-key"))
	g.Expect(profile.OAuthClientSecret).To(gomega.BeEmpty())
	g.Expect(validateProfileCredentials(profile, false)).ShouldNot(gomega.HaveOccurred())
}

func TestGetProfile_OAuthSecretEnvClearsFileJWT(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_TOKEN_ID", "")
	t.Setenv("MODAL_TOKEN_SECRET", "")
	t.Setenv("MODAL_OAUTH_CLIENT_SECRET", "ov-env-secret")
	t.Setenv("MODAL_OAUTH_JWT_KEY", "")

	profile := getProfile("oauth-profile", config{
		"oauth-profile": rawProfile{
			OAuthRefreshToken: "refresh-token",
			OAuthClientID:     "oc-client-id",
			OAuthJWTKey:       "file-jwt-key",
		},
	})
	g.Expect(profile.OAuthClientSecret).To(gomega.Equal("ov-env-secret"))
	g.Expect(profile.OAuthJWTKey).To(gomega.BeEmpty())
	g.Expect(validateProfileCredentials(profile, false)).ShouldNot(gomega.HaveOccurred())
}

func TestGetProfile_OAuthBothEnvMethodsKept(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_OAUTH_CLIENT_SECRET", "ov-env-secret")
	t.Setenv("MODAL_OAUTH_JWT_KEY", "env-jwt-key")

	profile := getProfile("", config{})
	g.Expect(profile.OAuthClientSecret).To(gomega.Equal("ov-env-secret"))
	g.Expect(profile.OAuthJWTKey).To(gomega.Equal("env-jwt-key"))
}

func TestGetProfile_OAuthCredentialsFromConfig(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_OAUTH_REFRESH_TOKEN", "")
	t.Setenv("MODAL_OAUTH_CLIENT_ID", "")
	t.Setenv("MODAL_OAUTH_CLIENT_SECRET", "")
	t.Setenv("MODAL_OAUTH_JWT_KEY", "")

	profile := getProfile("oauth-profile", config{
		"oauth-profile": rawProfile{
			OAuthRefreshToken: "refresh-token",
			OAuthClientID:     "oc-client-id",
			OAuthClientSecret: "ov-client-secret",
			OAuthJWTKey:       "private-key",
		},
	})
	g.Expect(profile.OAuthRefreshToken).To(gomega.Equal("refresh-token"))
	g.Expect(profile.OAuthClientID).To(gomega.Equal("oc-client-id"))
	g.Expect(profile.OAuthClientSecret).To(gomega.Equal("ov-client-secret"))
	g.Expect(profile.OAuthJWTKey).To(gomega.Equal("private-key"))
}

func TestGetProfile_SandboxV2Parsing(t *testing.T) {
	cases := []struct {
		envVal   string
		expected bool
	}{
		{"", false},
		{"0", false},
		{"false", false},
		{"False", false},
		{"1", true},
		{"true", true},
		{"yes", true},
	}

	for _, tc := range cases {
		t.Run("MODAL_SANDBOX_V2="+tc.envVal, func(t *testing.T) {
			g := gomega.NewWithT(t)
			t.Setenv("MODAL_SANDBOX_V2", tc.envVal)
			profile := getProfile("", config{})
			g.Expect(profile.SandboxV2).To(gomega.Equal(tc.expected))
		})
	}
}

func TestGetProfile_SandboxV2FromConfigFile(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_SANDBOX_V2", "")

	profile := getProfile("v2-profile", config{"v2-profile": rawProfile{SandboxV2: true}})
	g.Expect(profile.SandboxV2).To(gomega.BeTrue())
}

func TestGetProfile_SandboxV2EnvOverridesConfigFile(t *testing.T) {
	g := gomega.NewWithT(t)
	t.Setenv("MODAL_SANDBOX_V2", "0")

	profile := getProfile("v2-profile", config{"v2-profile": rawProfile{SandboxV2: true}})
	g.Expect(profile.SandboxV2).To(gomega.BeFalse())
}

func TestProfileIsLocalhost(t *testing.T) {
	g := gomega.NewWithT(t)
	p := Profile{ServerURL: "http://localhost:8889"}
	g.Expect(p.isLocalhost()).Should(gomega.BeTrue())
}

// A value that parses as a float is not necessarily a duration. Infinity, NaN
// and anything too large to hold must fall back rather than become a nonsense
// deadline.
func TestGetProfile_IdleTimeoutRejectsValuesThatAreNotDurations(t *testing.T) {
	defaults := getProfile("", config{})

	cases := []struct {
		envVar string
		want   time.Duration
	}{
		{"MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", defaults.SandboxChannelIdleTimeout},
	}
	bad := []string{"Inf", "+Inf", "Infinity", "-Inf", "NaN", "1e30", "-1", "nonsense"}

	for _, tc := range cases {
		for _, value := range bad {
			t.Run(tc.envVar+"="+value, func(t *testing.T) {
				g := gomega.NewWithT(t)
				t.Setenv(tc.envVar, value)

				profile := getProfile("", config{})
				got := profile.SandboxChannelIdleTimeout
				g.Expect(got).To(gomega.Equal(tc.want), "should have fallen back to the default")
			})
		}
	}
}

// A positive timeout too short to measure must not read as zero, which is how
// the release is turned off - that would invert what was asked for.
func TestGetProfile_IdleTimeoutKeepsAShortTimeoutPositive(t *testing.T) {
	g := gomega.NewWithT(t)

	t.Setenv("MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", "0.0000000001")

	profile := getProfile("", config{})
	g.Expect(profile.SandboxChannelIdleTimeout).To(gomega.BeNumerically(">", 0))
}

// The largest accepted value still converts to a positive duration. Go leaves
// an out-of-range float-to-int conversion to the platform, so a bound that
// allowed one would be negative on some and saturated on others.
func TestGetProfile_IdleTimeoutAtTheUpperBoundStaysPositive(t *testing.T) {
	g := gomega.NewWithT(t)

	t.Setenv("MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", strconv.FormatFloat(maxIdleTimeoutSeconds, 'f', -1, 64))

	profile := getProfile("", config{})
	g.Expect(profile.SandboxChannelIdleTimeout).To(gomega.BeNumerically(">", 0))
}

// Values that are durations still get through, zero included: it turns the
// release off rather than releasing at once.
func TestGetProfile_IdleTimeoutAcceptsSeconds(t *testing.T) {
	g := gomega.NewWithT(t)

	t.Setenv("MODAL_SANDBOX_CHANNEL_IDLE_TIMEOUT", "0")

	profile := getProfile("", config{})
	g.Expect(profile.SandboxChannelIdleTimeout).To(gomega.Equal(time.Duration(0)))
}
